src/wp-includes/block-template.php

```diff
@@ -228,7 +228,28 @@ function get_the_block_template_html() {
 	$content = $wp_embed->autoembed( $content );
 	$content = shortcode_unautop( $content );
 	$content = do_shortcode( $content );
-	$content = do_blocks( $content );
+
+	/*
+	 * Most block themes omit the `core/query` and `core/post-template` blocks in their singular content templates.
+	 * While this technically still works since singular content templates are always for only one post, it results in
+	 * the main query loop never being entered which causes bugs in core and the plugin ecosystem.
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates. The while loop will by
+	 * definition only go through a single iteration, i.e. `do_blocks()` is only called once.
+	 *
+	 * Even if the block template contained a `core/query` and `core/post-template` block referencing the main query
+	 * loop, it would not cause errors since it would use a cloned instance and go through the same loop of a single
+	 * post, within the actual main query loop.
+	 */
+	if ( is_singular() ) {
```

Comment:
@joemcgill Based on your other comment, what do you think about adding this? IMO it would make this code a bit safer, as it would ensure that, _if_ something weird is going on and the main query was tampered with in a way that would change the number of posts in the result, we would fall back to the old code (`else` clause).
```suggestion
	global $wp_query;
	if ( is_singular() && $wp_query->post_count && have_posts() ) {
```

Reply:
I do think this is probably safer, but pretty unclear what's going on here unless someone does a git blame and finds this conversation (👋🏻 hi future person, if so!). Can you add an inline comment explaining why this is here and why you need to check the `post_count`?

Reply:
@joemcgill Updated in https://github.com/WordPress/wordpress-develop/pull/5104/commits/766d453fbff73a99391f4f512126654f6e796bc2.